pkg/pillar/cmd/zedrouter/appnumonunet.go

```diff
@@ -0,0 +1,221 @@
+
+// Network Instance/underlay network IP Address Management,
+// App Number management Module
+
+package zedrouter
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+// Network Instance Application number/IP Address Management
+// ---------------------------------------------------------
+//  AppNum  Module (bitMap & persistent AppNumMap)
+//   - uNetAppNumMapGet()
+//   - uNetAppNumGet()
+//   - uNetAppNumAlloc()
+//   - uNetAppNumDelete()
+//   - networkInstanceAppNumMapInitialize()
+//   - networkInstanceAppNumMapDelete()
+//   - getDhcpAppNumCountOnDhcpRange()
+//   - getDhcpAppNumCountOnSubnet()
+//   - networkInstanceDhcpAppNumCountMax()
+//  BitMap : is for the Dynamic dhcp address/app number assignments
+// For BitMap:(DhcpRange.Start - DhcpRange.End)
+//    - bitMapAppNumAdd(), set a bit in the BitMap
+//    - bitMapAppNumDelete(), reset a bit in the BitMap
+
+// mapped on NetworkInstance ID
+var netInstAppNumMap map[string]*types.NetworkInstanceAppNumMapInfo
+
+// returns the AppNumMap structure, for a network instance
+func uNetAppNumMapGet(networkID string) *types.NetworkInstanceAppNumMapInfo {
+	if netInstAppNumMap != nil {
+		if appNumMapInfo, ok := netInstAppNumMap[networkID]; ok {
+			if appNumMapInfo.DhcpAppNumCount != 0 {
+				return appNumMapInfo
+			}
+		}
+	}
+	return nil
+}
+
+// Get the AppNum for a bridge or, an app instance with static IP
+func uNetAppNumGet(ctx *zedrouterContext, networkID string, appID string,
+	ipAddr net.IP, mac net.HardwareAddr) (int, error) {
+
+	appNumMapInfo := uNetAppNumMapGet(networkID)
+	if appNumMapInfo == nil || ipAddr == nil {
+		return types.AppNumInvalid, nil
+	}
+
+	// get from persistent storage
+	addrMap := persistentAddrMapGet(ctx, networkID, appID)
+	if addrMap != nil {
+		return addrMap.AppNumOnUNet, nil
+	}
+
+	// Dhcp Range Start
+	startIP := appNumMapInfo.StartIP.To4()
+	if startIP == nil {
+		errStr := fmt.Sprintf("dhcprange start address parse error")
+		return types.AppNumInvalid, errors.New(errStr)
+	}
+	start := uint(startIP[0])<<24 + uint(startIP[1])<<16 +
+		uint(startIP[2])<<8 + uint(startIP[3])
+	addrBytes := ipAddr.To4()
+	if addrBytes == nil {
+		return types.AppNumInvalid, errors.New("parse error")
+	}
+	addr := uint(addrBytes[0])<<24 + uint(addrBytes[1])<<16 +
+		uint(addrBytes[2])<<8 + uint(addrBytes[3])
+	appNum := int(addr - start)
+	return appNum, nil
+}
+
+// Get the AppNum for ephimeral underlay network entry app instance
+func uNetAppNumAlloc(ctx *zedrouterContext,
+	networkID string, appID string, mac net.HardwareAddr) (int, error) {
+
+	appNumMapInfo := uNetAppNumMapGet(networkID)
+	if appNumMapInfo == nil {
+		log.Functionf("uNetAppNumAlloc(%s): not Addressable\n",
+			networkID)
+		return types.AppNumInvalid, nil
+	}
+
+	// get from persistent storage
+	addrMap := persistentAddrMapGet(ctx, networkID, appID)
+	if addrMap != nil {
+		return addrMap.AppNumOnUNet, nil
+	}
+
+	// TBD:XXX, this loop needs to be optimized
```

Comment:
No it doesn't. In fact I think we should remove the bitmap since is it not hidden/abstracted from the callers.